### PR TITLE
Revamp runtime errors

### LINKED ISSUE #29244: Introduce Error Hierarchy for Ballerina Runtime Errors
**Description:**
$Subject
**Describe your task(s)**
Define error objects used by runtime in a single place, in a `.bal` file and introduce error inheritance for runtime errors. 
**Related Issues (optional):**

**Suggested Labels (optional):**
Team/jBallerina, Type/Task

**Suggested Assignees (optional):**
gabilang

### PR DESCRIPTION
**Pull request by gabilang**
Fixes  #29244

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/JsonUtils.java [hunk 1]
```diff
@@ -179,19 +176,22 @@ private static Object getMappingElement(Object json, BString elementName, boolea
                 return null;
             }
 
-            return ErrorCreator.createError(MAP_KEY_NOT_FOUND_ERROR, StringUtils
-                    .fromString("Key '" + elementName + "' not found in JSON mapping"));
+            return ErrorUtils.getRuntimeError(BALLERINA_LANG_MAP_PKG_ID,
```
Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/JsonUtils.java:179]: Can we format this to one line?

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/cli/CliProvider.java [hunk 2]
```diff
@@ -230,4 +234,18 @@ private CliArg getCliArg(Module module, VariableKey variableKey) {
...
+    private static int stringToByte(String value) {
+        int byteValue = Integer.parseInt(value);
+        if (!isByteLiteral(byteValue)) {
+            throw ErrorCreator.createError(StringUtils.fromString(String
+                    .format(INVALID_INT_TO_BYTE_CONVERSION, value)));
+        }
+        return byteValue;
+    }
+
+    static boolean isByteLiteral(long longValue) {
```
Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/cli/CliProvider.java:247]: Range check we should able to use RuntimeUtils.isByteLiteral

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/util/exceptions/BallerinaErrorReasons.java [hunk 3]
```diff
@@ -20,12 +20,7 @@
 import io.ballerina.runtime.api.utils.StringUtils;
```
Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/util/exceptions/BallerinaErrorReasons.java:20]: Do we still need any usages of this class?

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/util/exceptions/RuntimeErrors.java [hunk 4]
```diff
@@ -23,63 +23,15 @@
  */
 public enum RuntimeErrors {
```
Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/util/exceptions/RuntimeErrors.java:24]: Shall we merge RuntimeErrorType and RuntimeErrors class

File: langlib/lang.array/src/main/ballerina/errors.bal [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import ballerina/lang.'runtime as runtime_error;
```
Comment (warunalakshitha) [inline langlib/lang.array/src/main/ballerina/errors.bal:1]: missing licence heander

File: langlib/lang.boolean/src/main/ballerina/errors.bal [hunk 6]
```diff
@@ -0,0 +1,4 @@
+import ballerina/lang.runtime as runtime_error;
```
Comment (warunalakshitha) [inline langlib/lang.boolean/src/main/ballerina/errors.bal:1]: missing license header

File: langlib/lang.map/src/main/ballerina/errors.bal [hunk 7]
```diff
@@ -0,0 +1,15 @@
+
```
Comment (warunalakshitha) [inline langlib/lang.map/src/main/ballerina/errors.bal:1]: Shall we add the licence header here as well?

File: langlib/lang.string/src/main/ballerina/errors.bal [hunk 8]
```diff
@@ -0,0 +1,11 @@
+
```
Comment (warunalakshitha) [inline langlib/lang.string/src/main/ballerina/errors.bal:1]: Missing licence header. Let's check all other places too.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/JsonUtils.java (hunk 1 above)
Comment (gabilang) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/JsonUtils.java:179]: If I kept it in one line, the line length will exceed 120 characters (122 characters are there)

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/ErrorUtils.java [hunk 9]
```diff
@@ -99,38 +111,39 @@ public static ErrorValue createCancelledFutureError() {
...
+            throw ErrorCreator.createError(StringUtils.fromString(e.getMessage()));
+        }
     }
 
-    public static BError createNumericConversionError(Object inputValue, Type targetType) {
-        throw createError(BallerinaErrorReasons.NUMBER_CONVERSION_ERROR,
-                          BLangExceptionHelper.getErrorMessage(
-                                  RuntimeErrors.INCOMPATIBLE_SIMPLE_TYPE_CONVERT_OPERATION,
-                                  TypeChecker.getType(inputValue), inputValue, targetType));
+    public static BError getRuntimeError(Module module, RuntimeErrorType errorType, Object... params) {
```
Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/ErrorUtils.java:128]: Shall we change this method name to createRuntimeError for consistency with other methods?

File: langlib/lang.object/src/main/ballerina/errors.bal [hunk 10]
```diff
@@ -0,0 +1,8 @@
+import ballerina/lang.'runtime as runtime_error;
```
[the remark this answers is not in the document]
Comment (gabilang) [reply langlib/lang.object/src/main/ballerina/errors.bal:17]: added

File: langlib/lang.map/src/main/ballerina/errors.bal (hunk 7 above)
Comment (gabilang) [reply langlib/lang.map/src/main/ballerina/errors.bal:16]: added

File: langlib/lang.boolean/src/main/ballerina/errors.bal (hunk 6 above)
Comment (gabilang) [reply langlib/lang.boolean/src/main/ballerina/errors.bal:17]: added

File: langlib/lang.array/src/main/ballerina/errors.bal (hunk 5 above)
Comment (gabilang) [reply langlib/lang.array/src/main/ballerina/errors.bal:17]: added

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/cli/CliProvider.java (hunk 2 above)
Comment (gabilang) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/cli/CliProvider.java]: used accordingly

File: langlib/lang.string/src/main/ballerina/errors.bal (hunk 8 above)
Comment (gabilang) [reply langlib/lang.string/src/main/ballerina/errors.bal:16]: added

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/util/exceptions/BallerinaErrorReasons.java (hunk 3 above)
Comment (gabilang) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/util/exceptions/BallerinaErrorReasons.java:20]: Yes for now, there are few errors needed to be migrated to new error creator api.

**Pull request closed without merging**